### PR TITLE
Refactor module-level AST

### LINKED ISSUE #377: Use `ASTId` for names in AST
- [x] Constants,
- [x] fields,
- [x] functions,
- [x] The `traits` field of `ASTContract` and `ASTTrait`
- [x] and other AST elements

typically use `string` referring to the name of the element. It could be replaced by `ASTId`, so we'll get additional information about the definition point of the related AST element, which is useful for users of the compiler's API and internally to generate warnings referring to that element.

### LINKED ISSUE #498: Missing name information for error message on duplicate definitions from traits
```
trait Foo {
    receive("comment") { }
}

trait Bar {
    receive("comment") { }
}

contract Test with Foo, Bar { }
```

compilation fails (as it should) with the following (incorrect) error message:

```
Error: test.tact:11:1: Receive function for "[object Object]" already exists
Line 11, col 1:
  10 | 
> 11 | contract Test with Foo, Bar { }
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  12 | 

    at throwCompilationError (/Users/anton_1/ton/tact/dist/errors.js:64:11)
    at copyTraits (/Users/anton_1/ton/tact/dist/types/resolveDescriptors.js:1010:56)
    at processType (/Users/anton_1/ton/tact/dist/types/resolveDescriptors.js:1043:9)
    at processType (/Users/anton_1/ton/tact/dist/types/resolveDescriptors.js:1040:13)
    at processType (/Users/anton_1/ton/tact/dist/types/resolveDescriptors.js:1040:13)
    at resolveDescriptors (/Users/anton_1/ton/tact/dist/types/resolveDescriptors.js:1049:9)
    at precompile (/Users/anton_1/ton/tact/dist/pipeline/precompile.js:18:55)
    at build (/Users/anton_1/ton/tact/dist/pipeline/build.js:58:43)
    at Object.run (/Users/anton_1/ton/tact/dist/node.js:84:47)
```

the `copyTraits` function should not use `f.selector` for error reporting.

### LINKED ISSUE #501: Error message with weird JS/TS type coersion upon compiling a contract with a missing storage field declaration
Consider the following snippet:

```tact
contract Example {
    init() {
        self.nonExistentField = 42;
    }
}
```

Its compilation fails (as expected), but produces a weird error involving type coersion of JS/TS somewhere (not expected):

```shell
Error: example.tact:3:9: Type "Example" does not have a field named "[object Object]"
Line 3, col 9:
  2 |     init() {
> 3 |         self.nonExistentField = 42;
              ^~~~~~~~~~~~~~~~~~~~~
  4 |     }

    at throwCompilationError (node_modules/@tact-lang/compiler/dist/errors.js:64:11)
    at resolveFieldAccess (node_modules/@tact-lang/compiler/dist/types/resolveExpression.js:252:48)
    at resolveExpression (node_modules/@tact-lang/compiler/dist/types/resolveExpression.js:487:16)
    at processStatements (node_modules/@tact-lang/compiler/dist/types/resolveStatements.js:153:61)
    at processFunctionBody (node_modules/@tact-lang/compiler/dist/types/resolveStatements.js:391:17)
    at resolveStatements (node_modules/@tact-lang/compiler/dist/types/resolveStatements.js:446:19)
    at precompile (node_modules/@tact-lang/compiler/dist/pipeline/precompile.js:24:53)
    at build (node_modules/@tact-lang/compiler/dist/pipeline/build.js:58:43)
    at Object.run (node_modules/@tact-lang/compiler/dist/node.js:84:47)
```

### PR DESCRIPTION
**Pull request by anton-trunov**
Towards #518. Closes #377. Closes #498. Closes #501.

- [x] I have updated CHANGELOG.md
- ~[ ] I have documented my contribution in Tact Docs: https://github.com/tact-lang/tact-docs/pull/PR-NUMBER~
- [x] I have added tests to demonstrate the contribution is correctly implemented: this usually includes both positive and negative tests, showing the happy path(s) and featuring intentionally broken cases
- [x] I have run all the tests locally and no test failure was reported
- [x] I have run the linter, formatter and spellchecker
- [x] I did not do unrelated and/or undiscussed refactorings

### CODE REVIEW

File: src/grammar/ast.ts [hunk 1]
```diff
@@ -582,15 +633,20 @@ export type ASTExpression =
     | ASTInitOf
     | ASTString
     | ASTConditional;
-export type ASTType = ASTPrimitive | ASTStruct | ASTContract | ASTTrait;
+export type ASTType =
```
Comment (byakuren-hijiri) [inline src/grammar/ast.ts:636]: nit: Should it be `AstType`, for consistency?

File: src/grammar/ast.ts [hunk 2]
```diff
@@ -1,84 +1,238 @@
...
+            return left.text === right;
+        }
+        return left.text === right.text;
+    }
+}
+
+export const selfId: AstId = {
+    kind: "id",
+    text: "self",
+    id: 0,
```
Comment (byakuren-hijiri) [inline src/grammar/ast.ts:207]: Is it safe to use a `0` value here? I mean, could we have a clash if the parser generates a `0` ID elsewhere?

**Comment by byakuren-hijiri** [comment]
Looks massive!

The AST refactoring LGTM. I especially liked the separation between function declaration and definition, and the separate nodes for structures and messages.

I skimmed through changes that reflects to the AST modification, I did not find any changes in the intended behavior of the compiler.

File: src/grammar/ast.ts (hunk 1 above)
Comment (anton-trunov) [reply src/grammar/ast.ts:636]: yeah, it will be in a subsequent PR, this one only handles top-level things

File: src/grammar/ast.ts (hunk 2 above)
Comment (anton-trunov) [reply src/grammar/ast.ts:207]: Should be safe because we start with 1: `let nextId = 1`

**Comment by anton-trunov** [comment]
Thanks! I'll do one or two more PRs refactoring AST. Merging this one to so we can all sync

**Pull request merged**